src/realm/object-store/impl/object_notifier.cpp

```diff
@@ -32,22 +32,53 @@ ObjectNotifier::ObjectNotifier(std::shared_ptr<Realm> realm, TableKey table, Obj
 
 bool ObjectNotifier::do_add_required_change_info(TransactionChangeInfo& info)
 {
+    // The object might have been deleted in the meantime.
+    if (!m_table)
+        return false;
+
     m_info = &info;
-    info.tables[m_table.value];
-    return false;
+    info.tables[m_table->get_key().value];
+
+    // When adding or removing a callback the related tables can change due to the way we calculate related tables
+    // when key path filters are set hence we need to recalculate every time the callbacks are changed.
+    util::CheckedLockGuard lock(m_callback_mutex);
+    if (m_did_modify_callbacks) {
+        m_related_tables = {};
+        recalculate_key_path_arrays();
+        DeepChangeChecker::find_filtered_related_tables(m_related_tables, *m_table, m_key_path_arrays,
+                                                        all_callbacks_filtered());
+        m_did_modify_callbacks = false;
+    }
+
+    return true;
 }
 
 void ObjectNotifier::run()
 {
     if (!m_table)
         return;
 
-    auto it = m_info->tables.find(m_table.value);
+    if (!m_change.modifications.contains(0) && any_callbacks_filtered()) {
```

Comment:
As we discussed on a call:
This code path -- where a keypath is passed -- would need some kind of addition to column_modifications also (or at least some kind of why for the `Property` to be read from a realm::CollectionChangeSet).
Otherwise RLMObjectBase.after would not deliver the callback.

Reply:
Fixed with `a4442d5`.

Reply:
Nice. Also needs same behavior for deletions.
ex)
```
let token1 = object.observe(keyPaths: ["intCol"]) { change in
//...
}
realm.delete(object)
```
This would need have some sort of `m_change.deletions.add()` behavior.

Reply:
Fixed with e50dbc2.